### PR TITLE
Auto diffing section controller

### PR DESCRIPTION
**Pull request by rnystrom**
Started work on the plane to get this moving since #418  is up and ready to land. We'll likely need to spend some time fleshing out the API of this, and I think I'll split it up into a couple different PRs once ready for review. Putting this up now to get early feedback.

## What is this?

This adds an auto-diffing section controller as outlined in #38. There are several key parts:

- Subclass a new section controller `IGListAutoSectionController` (naming wip)
- Connect a data source
- Implement the data source methods that do 3 things:
  - Given a top-level object, transform it into an array of **diffable** view models
  - Given a view model, return a cell
  - Given a view model, return a size for a cell
- A new protocol for the cell `IGListBindable` so that we can control when the cell is updated w/ the view model.
  - The most important part of this is that it unlocks moving and reloading a cell, which you can't do w/ `UICollectionView`

## TODO

- [ ] Unit test `reloadObjects:`
- [x] Add `didSelectItemAtIndex:` support
- [x] Verify all headers use FB copyright
- [x] Create PR for `IGListAdapterUpdater` changes
  - [x] Include fix + tests for #288 
- [ ] ~~Create PR for re-entrant updates~~
  - ~~e.g. calling `performBatchAnimated:updates:completion:` within an in-progress update block should execute immediately~~ doing this here
- [x] Create PR for new section controller w/ unit tests
- [x] Create example
- [x] Add header docs
- [ ] Optimize test coverage

### IGListAdapter

I had to make some changes to `IGListAdapterUpdater` in order for this to work. The biggest challenge is when the section controller receives a new object, it has to split it up into view models and then apply changes.

However, `didUpdateToObject:` will be called _inside_ `performBatchUpdates:`, so we need to diff and fire off the changes on the collection view immediately. A few issues:

- You couldn't use `-[IGListCollectionContext performBatchAnimated:updates:completion:]` inside `didUpdateToObject:` and have it update immediately (async updates)
- There are [other ways](https://github.com/Instagram/IGListKit/blob/master/Source/IGListAdapter.m#L325) that `didUpdateToObject:` gets called, so we can't just insert/delete/move
- We're planning to remove non-batched updates entirely in #392

### Diffing top-level models

One super important piece to getting this right is having top-level (the ones given to `IGListAdapter` and handed to the SC in `didUpdateToObject:` always equal itself, even if its values change. I solved this in the unit tests w/ an object that implements diffing like:

```objc
@interface IGTestObject: NSObject <IGListDiffable>
@property NSString *key;
@end

@implementation IGTestObject
- (id)diffIdentifier {
  return self.key;
}

- (BOOL)isEqualToDiffableObject:(id)object {
  return [self.key isEqual:[object key]];
}
@end
```

If you use a traditional `isEqualToDiffableObject:` (where you check all props/values) you'll end up reloading the section, avoiding all the nice cell animations.

I started making a stock object that does this, something like `IGListAutoObject`, but its so simple that it's kind of annoying. It's essentially a model box that will probably end up more confusing than having to learn how to diff w/ the auto SC.

Would love thoughts here.

### Selection

`didSelectItemAtIndex:` is totally broken. Options are:

- Add a new method on the data source protocol
  - Gross b/c that's another empty method for a lot of folks
- Add another delegate protocol you can optionally wire up

We've discussed splitting off `didSelectItemAtIndex:` into a `selectionDelegate` (like the display/scroll/etc delegate) in #184 and elsewhere. That might help, though the view models array is private, so we'd have to expose that or create another delegate API.

## How can I help?

Since this is super early, please hold on nit picks, formatting, etc. I'll work on that as we go along. But there are some big things I'd love your thoughts on:

- The name `IGListAutoSectionController`. Anything better out there? I thought `IGListDiffableSectionController` but idk
- I'm using a data source to split the object into view models, return cells, and handle sizes. I started off passing the SC blocks on init that did the transform (sort of like `IGListSingleSectionController`). I thought the API was kind of gross though, and you have to subclass the SC anyways.
- `IGListBindable` is great, but you can't just use the view model to config the cell. For instance, how do you make the SC the cell delegate? That's why I'm leaving the method to still return a cell, but we will call bind before returning it. Is this ok?
- Sample project now or later?
- Let me know if you see some glaring Swift interop issues. I haven't tried it yet.
- Throw all the crazy test-case scenarios you got at me!

### CODE REVIEW

**Comment by heumn** [comment]
> The name IGListAutoSectionController. Anything better out there? I thought IGListDiffableSectionController but idk

I think it is a good name. It describes what it does pretty well without being too verbose 👌

**Comment by jessesquires** [comment]
How about `IGListAutoDiffSectionController` ?

**Comment by heumn** [comment]
Yeah, `IGListAutoDiffSectionController` is actually a better name

**Comment by rnystrom** [comment]
@jessesquires @heumn we had some convo in #449 with other ideas:

- IGListAutoSectionController
- IGListReactSectionController
- IGListDiffableSectionController @zhubofei 
- IGListDiffingSectionController @amonshiz
- IGList(Cell)BindingSectionController
- IGListViewModelSectionController

I like IGListAutoDiffSectionController too!

**Comment by rnystrom** [comment]
Will rebase off of #525 to lighten the load here.

**Comment by rnystrom** [comment]
Anyone got any ideas on what we could demo as an example? Something that isn't crazy complicated, but would show off cells updating automatically.

**Comment by heumn** [comment]
Maybe something like you do in the nested controller example, but use colours and circles to easier track the diffing inside the section? 

Can not wait to play with this :)

skjermbilde 2017-03-05 kl 13 03 17

**Comment by jessesquires** [comment]
@rnystrom - can we defer this to `3.1.0` or does this contain breaking changes?

**Comment by rnystrom** [comment]
@jessesquires no breaking changes, we could kick this from 3.0 but I think I'll have this working + example this weekend.

**Comment by rnystrom** [comment]
Ok, I'm getting super excited about this. Check out the demo (gif below).

I'm about ready to move this from WIP to review. Should be ready tomorrow.

diffing

**Comment by heumn** [comment]
Initial reaction to this is that you've just saved me hundreds of lines of code and a hack 😂  Looks great

I also postponed "Call mom" until the 25th, and rather "Started cycling" the 20th. No flashing reloading of the sections and things updated just as I wanted 👌 💯 

Maybe it could be an idea to add something like this to extend on how powerful this is?

call-mom


I also keep thinking how you could make it extra clear that the section models _has_ to follow this principle? No ideas apart from the obvious documentation that is needed :)


>One super important piece to getting this right is having top-level (the ones given to IGListAdapter and handed to the SC in didUpdateToObject: always equal itself, even if its values change.

**Comment by heumn** [comment]
Btw, I forgot to that big fucking thank you for the awesome work you are putting in @rnystrom 🙌

**Comment by rnystrom** [comment]
@heumn great idea! I'm going to add an assert inside `-[IGListDiffingSectionController didUpdateToObject:]` that checks `[object isEqual:self.object]` (e.g. two objects should always be equal otherwise a section-reload happens). Also going to load the headers w/ docs and examples.

And nice extension! Maybe we can add that as a follow up?

**Comment by heumn** [comment]
> @heumn great idea! I'm going to add an assert inside -[IGListDiffingSectionController didUpdateToObject:] that checks [object isEqual:self.object] (e.g. two objects should always be equal otherwise a section-reload happens). Also going to load the headers w/ docs and examples.

Perfect 🎉 

> And nice extension! Maybe we can add that as a follow up?

👌

**Comment by rnystrom** [comment]
Importing for internal review!

**Comment by jessesquires** [comment]
that calendar view though 😚 👌

**Comment by amonshiz** [comment]
Early morning review before a run is probably the best kind of review right?

**Comment by rnystrom** [comment]
@amonshiz forget to hit "Submit Review"? 🤔

File: Source/IGListDiffingSectionController.m [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+@property (nonatomic, strong) id object;
+@property (nonatomic, assign) IGListDiffingSectionState state;
+
+@end
+
+@implementation IGListDiffingSectionController
+
+#pragma mark - Public API
+
+- (void)updateAnimated:(BOOL)animated completion:(void (^)())completion {
```
Comment (rnystrom) [inline Source/IGListDiffingSectionController.m]: Good eye! Means I need to add a test for this too.

**Comment by rnystrom** [comment]
Waddup @iglistkit-bot catching my own mistakes. I'm in love w/ this tool.

**Comment by rnystrom** [comment]
I'm still going back and forth on naming. Now I'm starting to lean towards `IGListBindingSectionController` b/c:

- The section controller does the "binding"
- "Diffing" is an implementation detail

The emphasis of this API should be on the model -> view model -> binding parts, not that it diffs.

What do you all think? @jessesquires @heumn @PhilCai1993 @amonshiz

**Comment by amonshiz** [comment]
sure? i'm bad at naming.

**Comment by jessesquires** [comment]
i'll defer to @amonshiz for naming strategies.

andrewnamesthings

**Comment by heumn** [comment]
> I'm still going back and forth on naming. Now I'm starting to lean towards IGListBindingSectionController b/c:

> The section controller does the "binding"
> "Diffing" is an implementation detail
> The emphasis of this API should be on the model -> view model -> binding parts, not that it diffs.

> What do you all think? @jessesquires @heumn @PhilCai1993 @amonshiz

Good reasoning 👌  👍

**Comment by PhilCai1993** [comment]
It's excited! I wish I could download this commits but how? 😅

**Comment by heumn** [comment]
It was merged into master, so you can just point you podfile to the master branch :)

**Comment by jessesquires** [comment]
@PhilCai1993 check out installation guide

https://instagram.github.io/IGListKit/installation.html

**Comment by PhilCai1993** [comment]
May I have a question here? If in a collectionview with various "Month", since `isEqualToDiffableObject:` simply returns true, how to reload the entire collectionview?

**Comment by rnystrom** [comment]
@PhilCai1993 covering that in #621 I think

**Pull request closed without merging**